### PR TITLE
Refactor generation backend

### PR DESCRIPTION
**Pull request by StAlKeR7779**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [x] Feature
- [x] Bug Fix
- [x] Optimization
- [ ] Documentation Update
- [ ] Community Node Submission

## Have you discussed this change with the InvokeAI team?
- [x] Yes
- [ ] No, because:

      
## Have you updated all relevant documentation?
- [ ] Yes
- [x] No

## Description
- Remove SDXL raw prompt nodes
- SDXL and SD1/2 generation merged to same nodes - t2l/l2l
- Fixed - if no xformers installed we trying to enable attention slicing, ignoring torch-sdp availability
- Fixed - In SDXL negative prompt now creating zeroed tensor(according to official code)
- Added mask field to l2l node
- Removed inpaint node and all legacy code related to this node
- Pass info about seed in latents, so we can use it to initialize ancestral/sde schedulers
- t2l and l2l nodes moved from strength to denoising_start/end
- Removed code for noise threshold(@hipsterusername said that there no plans to restore this feature)
- Fixed - first preview image now not gray
- Fixed - report correct total step count in progress, added scheduler order in progress event
- Added MaskEdge and ColorCorrect nodes (@hipsterusername)

## Added/updated tests?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by keturn** [comment]
Can we merge #4175 first?

I admit it's partly my thirst for those sweet line-deletion credits, but also there are a half-dozen different things going on in a single PR here and if we can take advantage of an already-existing PR to cut it down in scope, I think that will make it easier to review.

File: invokeai/app/invocations/compel.py [hunk 1]
```diff
@@ -176,7 +157,7 @@ def _lora_loader():
 
 
 class SDXLPromptInvocationBase:
-    def run_clip_raw(self, context, clip_field, prompt, get_pooled, lora_prefix):
+    def run_clip_compel(self, context, clip_field, prompt, get_pooled, lora_prefix, zero_on_empty):
```
Comment (psychedelicious) [inline invokeai/app/invocations/compel.py:160]: like to see typings here

File: invokeai/app/invocations/image.py [hunk 2]
```diff
@@ -650,3 +652,168 @@ def invoke(self, context: InvocationContext) -> ImageOutput:
...
+        )
+
+        return MaskOutput(
+            mask=ImageField(image_name=image_dto.image_name),
+            width=image_dto.width,
+            height=image_dto.height,
+        )
+
+class ColorCorrectInvocation(BaseInvocation, PILInvocationConfig):
+
```
Comment (psychedelicious) [inline invokeai/app/invocations/image.py:706]:
```suggestion
"""Shifts the colors of a target image to match the reference image, optionally using a mask to only color-correct certain regions of the target image."""

```

File: invokeai/app/invocations/image.py [hunk 3]
```diff
@@ -650,3 +652,168 @@ def invoke(self, context: InvocationContext) -> ImageOutput:
...
+    type: Literal["color_correct"] = "color_correct"
+
+    init: Optional[ImageField] = Field(default=None, description="Initial image")
+    result: Optional[ImageField] = Field(default=None, description="Resulted image")
+    mask: Optional[ImageField] = Field(default=None, description="Mask image")
+    mask_blur_radius: float = Field(default=8, description="Mask blur radius")
```
Comment (psychedelicious) [inline invokeai/app/invocations/image.py:712]: names can be a bit clearer (needs changes throughout `invoke()`)
```suggestion
    reference: Optional[ImageField] = Field(default=None, description="Reference image for color-correction")
    target: Optional[ImageField] = Field(default=None, description="The image to color-correct")
    mask: Optional[ImageField] = Field(default=None, description="Mask to use when applying color-correction")
    mask_blur_radius: float = Field(default=8, description="Mask blur radius")
```

File: invokeai/app/invocations/latent.py [hunk 4]
```diff
@@ -385,8 +423,16 @@ class LatentsToLatentsInvocation(TextToLatentsInvocation):
     type: Literal["l2l"] = "l2l"
 
     # Inputs
+    noise: Optional[LatentsField] = Field(description="The noise to use (test override for future optional)")
+
+    denoising_start: float = Field(default=0.0, ge=0, le=1, description="")
+    #denoising_end: float = Field(default=1.0, ge=0, le=1, description="")
```
Comment (psychedelicious) [inline invokeai/app/invocations/latent.py:429]: Now that `strength` is a userland concept:
- We need a formulas for `strength` <-> `denoising_start` implemented in JS. I know they are very simple, just making a note.
- We need to apply those formulas not only when creating graphs, but also when recalling metadata
- Do we change metadata to only have `denoising_start`? I think so

File: invokeai/app/invocations/latent.py [hunk 5]
```diff
@@ -198,7 +206,7 @@ def get_conditioning_data(
             # for ddim scheduler
             eta=0.0,  # ddim_eta
             # for ancestral and sde schedulers
-            generator=torch.Generator(device=unet.device).manual_seed(0),
+            generator=torch.Generator(device=unet.device).manual_seed(seed),
```
Comment (psychedelicious) [inline invokeai/app/invocations/latent.py:209]: @keturn `LatentsField` now optionally includes the `seed` used for noise generation, so we get deterministic-but-not-same-y generations on SDE & Ancestral samplers

File: invokeai/app/invocations/latent.py [hunk 6]
```diff
@@ -310,17 +318,37 @@ def prep_control_data(
...
+        timesteps = timesteps[: num_warmup_steps + scheduler.order * num_inference_steps]
+
+        return num_inference_steps, timesteps
+
+
     @torch.no_grad()
     def invoke(self, context: InvocationContext) -> LatentsOutput:
         with SilenceWarnings():
             noise = context.services.latents.get(self.noise.latents_name)
+            seed = self.noise.seed or 0
```
Comment (psychedelicious) [inline invokeai/app/invocations/latent.py:344]: This works but is a bit unintuitive - if `seed == 0`, we end up using the right-hand side to set it to `0` again.

Prefer checking for `None` and assigning to zero, even though functionally it's identical

File: invokeai/app/invocations/latent.py [hunk 7]
```diff
@@ -360,11 +388,21 @@ def _lora_loader():
                     exit_stack=exit_stack,
                 )
 
+                num_inference_steps, timesteps = self.init_scheduler(
+                    scheduler,
+                    device=unet.device,
+                    steps=self.steps,
+                    denoising_start=0.0,
+                    denoising_end=self.denoising_end,
+                )
```
Comment (psychedelicious) [inline invokeai/app/invocations/latent.py:397]: It seems like `t2l` and `l2l` can almost be merged? If I understand correctly, the differences are:
- `t2l` starts with full-noise while `l2l` starts with partial-noise
- `t2l` always denoises from 0, while `l2l` denoises from user-defined value

What else is different? Do we even need two nodes?

File: invokeai/app/invocations/latent.py [hunk 8]
```diff
@@ -385,8 +423,16 @@ class LatentsToLatentsInvocation(TextToLatentsInvocation):
...
+
+    denoising_start: float = Field(default=0.0, ge=0, le=1, description="")
+    #denoising_end: float = Field(default=1.0, ge=0, le=1, description="")
+
     latents: Optional[LatentsField] = Field(description="The latents to use as a base image")
-    strength: float = Field(default=0.7, ge=0, le=1, description="The strength of the latents to use")
+
+    mask: Optional[ImageField] = Field(
+        None, description="Mask",
+    )
```
Comment (psychedelicious) [inline invokeai/app/invocations/latent.py:435]: So when we think of `inpaint`, we are thinking of a "workflow". And our `l2l` logic optional accepts a mask.

What do you get when you `l2l + mask`, but without the rest of the `inpaint` workflow? Just a crappy inpaint?

File: invokeai/app/invocations/latent.py [hunk 9]
```diff
@@ -402,18 +448,42 @@ class Config(InvocationConfig):
...
+        mask_image = context.services.images.get_pil_image(mask.image_name)
+        if mask_image.mode != "L":
+            # FIXME: why do we get passed an RGB image here? We can only use single-channel.
+            mask_image = mask_image.convert("L")
+        mask_tensor = image_resized_to_grid_as_tensor(mask_image, normalize=False)
+        if mask_tensor.dim() == 3:
+            mask_tensor = mask_tensor.unsqueeze(0)
+        mask_tensor = tv_resize(
+            mask_tensor, lantents.shape[-2:], T.InterpolationMode.BILINEAR
+        )
```
Comment (psychedelicious) [inline invokeai/app/invocations/latent.py:464]: Are we sure that bilinear is the best option? Since this is just scaling the mask down, I guess it doesn't *really* matter.

File: invokeai/app/invocations/latent.py [hunk 10]
```diff
@@ -402,18 +448,42 @@ class Config(InvocationConfig):
...
+        return 1 - mask_tensor
+
     @torch.no_grad()
     def invoke(self, context: InvocationContext) -> LatentsOutput:
         with SilenceWarnings():  # this quenches NSFW nag from diffusers
-            noise = context.services.latents.get(self.noise.latents_name)
             latent = context.services.latents.get(self.latents.latents_name)
+            seed = self.latents.seed or 0
```
Comment (psychedelicious) [inline invokeai/app/invocations/latent.py:471]: same comment as in `t2l`

File: invokeai/app/services/events.py [hunk 11]
```diff
@@ -46,6 +47,7 @@ def emit_generator_progress(
                 source_node_id=source_node_id,
                 progress_image=progress_image.dict() if progress_image is not None else None,
                 step=step,
+                order=order,
```
Comment (psychedelicious) [inline invokeai/app/services/events.py:50]: Can you write the formula for this here in the PR please? for reference when updating frontendd

File: invokeai/app/util/step_callback.py [hunk 12]
```diff
@@ -56,23 +57,51 @@ def stable_diffusion_step_callback(
...
+            ],
+            dtype=sample.dtype,
+            device=sample.device,
+        )
+
+        sdxl_smooth_matrix = torch.tensor(
+            [
+                # [ 0.0478,  0.1285,  0.0478],
+                # [ 0.1285,  0.2948,  0.1285],
+                # [ 0.0478,  0.1285,  0.0478],
```
Comment (psychedelicious) [inline invokeai/app/util/step_callback.py:77]: Is this the matrix from another project? Maybe best to remove depending on license

Comment (psychedelicious) [reply invokeai/app/util/step_callback.py:77]: Also give yourself credit for experimentally determining your own!

**Review by psychedelicious — CHANGES_REQUESTED** [review]
🔥

File: invokeai/app/services/events.py (hunk 11 above)
Comment (StAlKeR7779) [reply invokeai/app/services/events.py:50]: To be 100% accurate it's:
```python
if order == 2:
    step = (step + 1) // 2
    total_steps = (total_steps + 1) // 2
```
For example Heun 20 steps - will report total_steps=39
In theory there could be schedulers with higher order, but there no such schedulers in diffusers currently.

File: invokeai/app/util/step_callback.py (hunk 12 above)
Comment (StAlKeR7779) [reply invokeai/app/util/step_callback.py:77]: Ok, i'll add my credits)
This just filter matrix before I reduced it's strength.
So, one that bellow based on this(multiplicated by 0.75 and recalculated center value)

File: invokeai/app/services/events.py (hunk 11 above)
Comment (psychedelicious) [reply invokeai/app/services/events.py:50]: To generalize:
```py
step = (step + 1) // order
total_steps = (total_steps + 1) // order
```

Or is `2` unrelated to `order`?

Are the valid `order`s currently `1 | 2`?

File: invokeai/app/invocations/latent.py (hunk 9 above)
Comment (StAlKeR7779) [reply invokeai/app/invocations/latent.py:464]: If you know what better I open to suggestions)
This code - copy from old inpaint generator code.

File: invokeai/app/invocations/latent.py (hunk 8 above)
Comment (StAlKeR7779) [reply invokeai/app/invocations/latent.py:435]: Yes, but in reality - it's not that crappy as denoising take in account, original surrounding pixels.
For example you can look at this picture. It's l2l + mask on normal image with anime model, so completely incompatible, but you can't see sharp border.
2361ffe1-d63f-4808-b181-6c649e0ed03f

File: invokeai/app/invocations/latent.py (hunk 6 above)
Comment (StAlKeR7779) [reply invokeai/app/invocations/latent.py:344]: Ok... I thought it's not uncommon patter, but I'll change to if

File: invokeai/app/services/events.py (hunk 11 above)
Comment (StAlKeR7779) [reply invokeai/app/services/events.py:50]: Yes, currently there exists only 1 and 2(heun, kdpm2, dpmpp_sde).
I not sure how it's correct to generalize:
```python
if order > 1:
    step = (step + 1) // order
    total_steps = (total_steps + 1) // order
```
or
```python
if order > 1:
    step = (step + order - 1) // order
    total_steps = (total_steps + order - 1) // order
```
So... that's why I wrote only about order=2)

File: invokeai/app/invocations/latent.py (hunk 5 above)
Comment (keturn) [reply invokeai/app/invocations/latent.py:205]: hmm. If `seed` here is the same as was used for the generator that generated the initial noise, is this going to re-create that initial noise its first time through the scheduler? That's probably not what we want.

In which case... well, the simplest thing would be to throw in some static adjustment here, like
```py
SCHEDULER_SALT = 2023080917  # some constant we make up, with enough bits to be interesting
scheduler_seed = initial_seed ^ SCHEDULER_SALT
ConditioningData(generator=Generator().seed(scheduler_seed))
```

another alternative might be to mirror the diffusers API and pass around the Generator itself instead of seeds... but the Generator is stateful so if the order that nodes execute in is not stable, that would prevent things from being reproducible.

File: invokeai/app/invocations/latent.py (hunk 7 above)
Comment (keturn) [reply invokeai/app/invocations/latent.py:392]: Thanks for bringing this up!

You are correct, these are not different operations.

It gets `latents` as input, and sometimes those latents will be right straight out of the Gaussian noise generator, and sometimes they'll be from something else. So there isn't need for a separate `noise` input (unless that's a Generator or something).

File: invokeai/app/invocations/latent.py (hunk 9 above)
Comment (keturn) [reply invokeai/app/invocations/latent.py:459]: I'm not at all sure, but yeah, if this is scaling it down to 1/8th size, I think it's going to be hard to notice a difference here.

File: invokeai/backend/stable_diffusion/diffusers_pipeline.py [hunk 13]
```diff
@@ -127,11 +123,10 @@ def apply_mask(self, latents: torch.Tensor, t) -> torch.Tensor:
...
             # TODO: file diffusers bug about inconsistency?
             t = einops.repeat(t, "-> batch", batch=batch_size)
-        # Noise shouldn't be re-randomized between steps here. The multistep schedulers
-        # get very confused about what is happening from step to step when we do that.
-        mask_latents = self.scheduler.add_noise(self.mask_latents, self.noise, t)
-        # TODO: Do we need to also apply scheduler.scale_model_input? Or is add_noise appropriately scaled already?
-        # mask_latents = self.scheduler.scale_model_input(mask_latents, t)
+
+        if self.noise is not None:
+            mask_latents = self.scheduler.add_noise(self.mask_latents, self.noise, t)
```
Comment (keturn) [inline invokeai/backend/stable_diffusion/diffusers_pipeline.py:128]: Why did `noise` become optional here?

I think I had this always on quite deliberately. If the overall noisiness of the input does not match the level expected for _t_, the model does not handle it well.

File: invokeai/backend/stable_diffusion/diffusers_pipeline.py [hunk 14]
```diff
@@ -207,8 +202,8 @@ class ControlNetData:
 
 @dataclass
 class ConditioningData:
-    unconditioned_embeddings: torch.Tensor
-    text_embeddings: torch.Tensor
+    unconditioned_embeddings: Any # TODO: type
+    text_embeddings: Any # TODO: type
```
Comment (keturn) [inline invokeai/backend/stable_diffusion/diffusers_pipeline.py:206]: Are these not tensors anymore?

File: invokeai/app/services/events.py (hunk 11 above)
Comment (keturn) [reply invokeai/app/services/events.py:50]: We shouldn't have to reverse-engineer this; the thing emitting the progress events has access to the complete list of timesteps it's planning to run.

**Review by keturn — COMMENTED** [review]
Thanks for merging in the other code removal first!

That knocked down the size of this PR a bit, but IMHO it's still too much for a single PR.

There's inpainting stuff that could be done without the SDXL stuff,  
the seeding of the stochastic schedulers is an independent issue,  
the changes to adjust-memory-efficient-attention could and should be considered and tested on their own--

I've left a few comments so far, but this does not constitute anything near a complete review. I haven't even got to the changes in `shared_invokeai_diffusion` yet, let alone any of the SDXL.

File: invokeai/app/services/events.py (hunk 11 above)
Comment (StAlKeR7779) [reply invokeai/app/services/events.py:50]: I send raw values because we can show progress more smooth, than when we will skip every second unet run on order=2 schedulers.
Sended before code with `(v + 1) // 2` already output correct values.
Or if you about schedulers internals - I read them, but here no comments why it's done this way.
I understand why it repeat each timestep twice(because they do different calculations on each step), but why for example first timestep not repeated too - there no comments.

File: invokeai/backend/stable_diffusion/diffusers_pipeline.py (hunk 14 above)
Comment (StAlKeR7779) [reply invokeai/backend/stable_diffusion/diffusers_pipeline.py]: Yes, now it's conditionings object with all info about this conditionings.

File: invokeai/app/invocations/latent.py (hunk 5 above)
Comment (StAlKeR7779) [reply invokeai/app/invocations/latent.py]: Added xor by 0xFFFFFFFF to flip all bits.

**Review by psychedelicious — APPROVED** [review]
